### ISSUE TITLE
[react/prop-types] False positive when wrapped with React.memo

### ISSUE THREAD

**Issue by bhollis** [yannickcr/eslint-plugin-react#2760]
Using version 7.20.6.

I have a TypeScript component like:

```tsx
export default React.memo(function AppIcon({
  icon,
  className,
  title,
  spinning,
}: {
  icon: string | IconDefinition;
  className?: string;
  title?: string;
  spinning?: boolean;
}) {
  if (typeof icon === 'string') {
    return (
      <span
        className={clsx(
          icon,
          'app-icon',
          'no-pointer-events',
          className,
          spinning ? 'fa-spin' : false
        )}
        title={title}
      />
    );
  } else {
    return (
      <FontAwesomeIcon
        className={className ? 'app-icon ' + className : 'app-icon'}
        icon={icon}
        title={title}
        spin={spinning}
      />
    );
  }
});
```

This falls afoul of `react/prop-types`:

```
  11:3  error  'icon' is missing in props validation       react/prop-types
  12:3  error  'className' is missing in props validation  react/prop-types
  13:3  error  'title' is missing in props validation      react/prop-types
  14:3  error  'spinning' is missing in props validation   react/prop-types
```

However if I move the function declaration out of the `React.memo` call and then just have `export default React.memo(AppIcon)`, the warning goes away.

**Comment by ljharb** [comment]
I assume that if the warning goes away, it's because you have `AppIcon.propTypes =` somewhere?

**Comment by ljharb** [comment]
ah, this is typescript, so it's using the types. Indeed, this seems like a bug in prop types detection.

**Comment by dididy** [comment]
I'll take a look this issue.

**Comment by fairbanksg** [comment]
I am having a similar problem. Strangely, I do not get false positives when just using memo, but when combining memo and forwardRef.

Also using Typescript:
```tsx
type FooProps = { x: string; y: number };
const Foo = memo(
  forwardRef<HTMLDivElement, FooProps>(function Foo({ x, y }: FooProps, ref) {
    return (
      <div ref={ref}>
        {x} {y}
      </div>
    );
  })
);
```

I get 
```
'x' is missing in props validation eslint(react/prop-types)
'y' is missing in props validation eslint(react/prop-types)
```

**Comment by ranquild** [comment]
Use can use the following workaround:

```
export const memo: <T>(c: T) => T = React.memo;
```

So the props type will be correctly preserved.

**Comment by tamakisquare** [comment]
I have the same issue but with using only `forwardRef`.

**Comment by httpete** [comment]
Same issue with latest TypeScript 4.4.3 and forwardRef as @fairbanksg

**Comment by dgattey** [comment]
Thanks for a useful rule! I was wondering about the status of a fix for this. It's still a bug with React 17.0.2's `forwardRef` & Typescript 4.3.5. This minimal example reproduces it (see the screenshot):
```typescript
const IndeterminateCheckbox = React.forwardRef<
  HTMLInputElement,
  React.ComponentPropsWithoutRef<'input'>
>((props, ref) => {
  const { checked } = props;
  return <input ref={ref} checked={checked} />
});
```
Worth noting that the type _is_ found and is correct, but the rule still fails 
![image](https://user-images.githubusercontent.com/982182/144159163-815faf23-1d6a-49d2-a7ff-6bbe03d41ad8.png)

**Comment by pioz** [comment]
I have a similar issue with this code:
```jsx
import React, { forwardRef } from 'react'

export interface ScannerProps {
  width: number
  className?: string
  deviceId: string | null
  mirror: boolean
  xAdjustment: number
  yAdjustment: number
  zoomAdjustment: number
  loading: boolean
  onClick: (() => void) | undefined
  children: React.ReactChild | React.ReactChildren | Array<React.ReactChild> | Array<React.ReactChildren> | null
}

const Scanner = forwardRef<HTMLCanvasElement, ScannerProps>(
  (
    {
      width,
      className,
      deviceId,
      mirror,
      xAdjustment,
      yAdjustment,
      zoomAdjustment,
      loading,
      onClick,
      children
    }: ScannerProps,
    ref: React.ForwardedRef<HTMLCanvasElement>
  ): JSX.Element => { ... }

export default Scanner
```
I get the following eslint errors:
```
'width' is missing in props validation eslint(react/prop-types)
'className' is missing in props validation eslint(react/prop-types)
'deviceId' is missing in props validation eslint(react/prop-types)
...
'children' is missing in props validation eslint(react/prop-types)
```

Any idea?

**Comment by ljharb** [comment]
The status is, just like anything else on GitHub, “no update, or else it’d be posted on the issue”.

If you’d like to submit a PR that fixes it - or even that just has failing test cases - that’d help move things along.

**Comment by ChazUK** [comment]
I fixed this using the following

```
interface Props {
  name: string
  index: number
}

const SomeComponent: FC<Props> = ({
  name,
  index
}) => {
  ...
}

export default memo(SomeComponent);
```

**Comment by vaynevayne** [comment]
> I fixed this using the following
> 
> ```
> interface Props {
>   name: string
>   index: number
> }
> 
> const SomeComponent: FC<Props> = ({
>   name,
>   index
> }) => {
>   ...
> }
> 
> export default memo(SomeComponent);
> ```
not success in vite4 , cause error,  It roughly means : react host refresh need unique export

**Comment by JordanDC2** [comment]
this is still an issue in the latest version of eslint & eslint-plugin-react, it would be nice to reopen this issue

**Comment by ljharb** [comment]
If there's still a problem with v7.33.0, please file a new issue.

**Comment by obf1313** [comment]
i have same problem, can anybody help me?
